### ISSUE TITLE
ALS is deprecated and archived, Better to use AsyncLocalStorage class fom the async_hook std module

### ISSUE THREAD

**Issue by anks333** [fastify/fastify-request-context#148]
https://github.com/vicanso/async-local-storage is archived.

Example:

```ts
// ctx.ts
import { AsyncLocalStorage } from "async_hooks";

export const asyncLocalStorage = new AsyncLocalStorage<Map<string, unknown>>();

export const setInCtx = <T = unknown>(key: string, value: T): void => {
  asyncLocalStorage.getStore()?.set(key, value);
}

export const getFromCtx = <T = unknown>(key: string): T | undefined => asyncLocalStorage.getStore()?.get(key) as T;

export const initReqContext = (req: any, _res: any, next: any) => {
  asyncLocalStorage.run(new Map(), () => {
    const store = asyncLocalStorage.getStore();
    if (store){
      store.set("key",  "value");
    }
    next();
  });
}
```

```ts
// server.ts / index.ts 

import { initReqContext } from './ctx';

// this only works with preHandler hooks
server.addHook("preHandler", initReqContext)

```

you can start using get and set function to access the store

**Comment by kibertoad** [comment]
Check out the underlying library that we use, it already uses ALS when supported by used Node version.

**Comment by Uzlopak** [comment]
As far as I can see if is part of node 14. So i agree, we can replace this dependency with the built in.

Can you provide a PR?

**Comment by anks333** [comment]
This is not an issue, but I struggled to integrate the `AsyncLocalStorage ` with fastify so I am sharing the solution here.
I noticed it only work with the `preHandler` hook.

**Comment by kibertoad** [comment]
This plugin works with any lifecycle phase, and does not use the library that you mentioned.
@Uzlopak what's the minimum Node requirement for fastify 4?

**Comment by Uzlopak** [comment]
node 14 is the minimum requirement.

**Comment by kibertoad** [comment]
I will create a PR then

**Comment by voxpelli** [comment]
I think there's a mix up between @kibertoad's [`asynchronous-local-storage`](https://www.npmjs.com/package/asynchronous-local-storage) module which is used here and the archived [`async-local-storage`](https://github.com/vicanso/async-local-storage) module.

The `asynchronous-local-storage` module uses the `AsyncLocalStorage` from `node:async_hooks` when that's enabled, else downgrades to [`cls-hooked`](https://www.npmjs.com/package/cls-hooked).

It's probably time to drop `cls-hooked` in `asynchronous-local-storage`, but that's a separate issue for that repository I think, its only an issue for this project if that would make `asynchronous-local-storage` so small that it would make more sense to include it directly within this module.

**Comment by kibertoad** [comment]
Yeah, I was planning to do that for quite a while. PR would be welcome

**Comment by voxpelli** [comment]
> Yeah, I was planning to do that for quite a while. PR would be welcome

@kibertoad PR for which of the two?

1. Removing `cls-hooked` from `asynchronous-local-storage`
2. Moving the `node:async_hooks` based part of `asynchronous-local-storage` into this module, removing `asynchronous-local-storage` as a separate dependency

**Comment by kibertoad** [comment]
The second one.

**Comment by mcollina** [comment]
While bumping Fastify major is _hard_, bumping an individual package is easy. There are good reasons to bump this package major version and drop Node.js v14.

**Comment by voxpelli** [comment]
I'm making a PR
